crates/prover/src/proving.rs

```diff
@@ -0,0 +1,403 @@
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use borsh::{BorshDeserialize, BorshSerialize};
+use citrea_common::cache::L1BlockCache;
+use citrea_common::da::extract_sorted_sequencer_commitments;
+use citrea_common::utils::{check_l2_range_exists, filter_out_proven_commitments};
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use sov_db::ledger_db::ProverLedgerOps;
+use sov_db::schema::types::{BatchNumber, StoredProof, StoredStateTransition};
+use sov_modules_api::{BlobReaderTrait, SlotData, SpecId, Zkvm};
+use sov_rollup_interface::da::{BlockHeaderTrait, DaSpec, SequencerCommitment};
+use sov_rollup_interface::rpc::SoftConfirmationStatus;
+use sov_rollup_interface::services::da::DaService;
+use sov_rollup_interface::zk::{Proof, StateTransitionData, ZkvmHost};
+use sov_stf_runner::ProverService;
+use tokio::sync::Mutex;
+use tracing::{error, info, warn};
+
+use crate::da_block_handler::{
+    break_sequencer_commitments_into_groups, get_state_transition_data_from_commitments,
+};
+use crate::errors::L1ProcessingError;
+
+type TxId<Da> = <Da as DaService>::TransactionId;
+
+pub(crate) async fn data_to_prove<Da, DB, StateRoot, Witness>(
+    da_service: Arc<Da>,
+    ledger: DB,
+    sequencer_pub_key: Vec<u8>,
+    sequencer_da_pub_key: Vec<u8>,
+    l1_block_cache: Arc<Mutex<L1BlockCache<Da>>>,
```

Comment:
`l1_block_cache` goes into `get_da_block_at_height` only. So we can add `l1_block_cache` to BitcoinDa and add a common method for DaService trait so we don't have to pass that block_cache around.

Reply:
well actually it can be in the da adapter implementations, however, that's a seperate issue imo